runners/flink/runner/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/io/UnboundedSourceWrapper.java

```diff
@@ -18,107 +18,255 @@
+      int parallelism) throws Exception {
+    this.serializedOptions = new SerializedPipelineOptions(pipelineOptions);
+
+    if (source.requiresDeduping()) {
+      LOG.warn("Source {} requires deduping but Flink runner doesn't support this yet.", source);
     }
-    return WindowedValue.of(output, timestamp, GlobalWindow.INSTANCE, PaneInfo.NO_FIRING);
+
+    checkpointMarkCoder = source.getCheckpointMarkCoder();
+    // get the splits early. we assume that the generated splits are stable,
+    // this is necessary so that the mapping of state to source is correct
+    // when restoring
+    splitSources = source.generateInitialSplits(parallelism, pipelineOptions);
   }
 
   @Override
-  public void run(SourceContext<WindowedValue<T>> ctx) throws Exception {
+  public void run(SourceContext<WindowedValue<OutputT>> ctx) throws Exception {
     if (!(ctx instanceof StreamSource.ManualWatermarkContext)) {
       throw new RuntimeException(
-          "We assume that all sources in Dataflow are EventTimeSourceFunction. " +
-              "Apparently " + this.name + " is not. " +
-              "Probably you should consider writing your own Wrapper for this source.");
+          "Cannot emit watermarks, this hints at a misconfiguration/bug.");
     }
 
-    context = (StreamSource.ManualWatermarkContext<WindowedValue<T>>) ctx;
-    runtime = (StreamingRuntimeContext) getRuntimeContext();
+    context = (StreamSource.ManualWatermarkContext<WindowedValue<OutputT>>) ctx;
+    runtimeContext = (StreamingRuntimeContext) getRuntimeContext();
+
+    // figure out which split sources we're responsible for
+    int subtaskIndex = getRuntimeContext().getIndexOfThisSubtask();
+    int numSubtasks = getRuntimeContext().getNumberOfParallelSubtasks();
 
-    isRunning = true;
+    List<UnboundedSource<OutputT, CheckpointMarkT>> localSources = new ArrayList<>();
 
-    reader = source.createReader(serializedOptions.getPipelineOptions(), null);
+    for (int i = 0; i < splitSources.size(); i++) {
+      if (i % numSubtasks == subtaskIndex) {
+        localSources.add(splitSources.get(i));
+      }
+    }
 
-    boolean inputAvailable = reader.start();
+    LOG.info("Unbounded Flink Source {}/{} is reading from sources: {}",
+        subtaskIndex,
+        numSubtasks,
+        localSources);
+
+    readers = new ArrayList<>();
+    if (restoredState != null) {
+      // we are restoring from a checkpoint, initialize the readers with the
+      // CheckpointMark. we assume a consistent order in the sources/readers,
+      // so that the order of the CheckpointMarks matches the order of
+      // our readers
+      try (ByteArrayInputStream bais = new ByteArrayInputStream(restoredState)) {
+        for (UnboundedSource<OutputT, CheckpointMarkT> source : localSources) {
+          CheckpointMarkT checkpointMark = checkpointMarkCoder.decode(bais, Coder.Context.OUTER);
+          readers.add(source.createReader(serializedOptions.getPipelineOptions(), checkpointMark));
```

Comment:
the `OUTER` context here does not seem right. This implies that each checkpoint owns the entire stream. I would think you would probably use a `ListCoder<CheckpointMarkT>` and just decode that list. That way the individual marks would be delimited properly, etc.